### PR TITLE
Simplify data conversion from Spark to PyTorch DataLoader

### PR DESCRIPTION
**Pull request by liangz1**
# What changes are proposed in this PR?
Add `converter.make_torch_dataloader()` with advanced params.

# The latest API
```python
def make_spark_converter(
        df,
        parquet_row_group_size_bytes=DEFAULT_ROW_GROUP_SIZE_BYTES,
        compression_codec=None):
    """
    Convert a spark dataframe into a :class:`SparkDatasetConverter` object.
    It will materialize a spark dataframe to the directory specified by
    spark conf 'petastorm.spark.converter.parentCacheDirUrl'.
    The dataframe will be materialized in parquet format, and we can specify
    `parquet_row_group_size_bytes` and `compression_codec` for the parquet
    format. See params documentation for details.

    The returned `SparkDatasetConverter` object will hold the materialized
    dataframe, and can be used to make one or more tensorflow datasets or
    torch dataloaders.

    We can explicitly delete the materialized dataframe data, see
    `SparkDatasetConverter.delete`, and when the spark application exit,
    it will try best effort to delete the materialized dataframe data.

    :param df: The :class:`DataFrame` object to be converted.
    :param parquet_row_group_size_bytes: An int denoting the number of bytes
        in a parquet row group when materializing the dataframe.
    :param compression_codec: Specify compression codec.
        It can be one of 'uncompressed', 'bzip2', 'gzip', 'lz4', 'snappy', 'deflate'.
        Default None. If None, it will leave the data uncompressed.

    :return: a :class:`SparkDatasetConverter` object that holds the
        materialized dataframe and can be used to make one or more tensorflow
        datasets or torch dataloaders.
    """

class SparkDatasetConverter(object):
    """
    A `SparkDatasetConverter` object holds one materialized spark dataframe and
    can be used to make one or more tensorflow datasets or torch dataloaders.
    The `SparkDatasetConverter` object is picklable and can be used in remote
    processes.
    See `make_spark_converter`
    """

    PARENT_CACHE_DIR_URL_CONF = 'petastorm.spark.converter.parentCacheDirUrl'

    def __init__(self, cache_dir_url, dataset_size):
        """
        :param cache_dir_url: A string denoting the path to store the cache
            files.
        :param dataset_size: An int denoting the number of rows in the
            dataframe.
        """

    def __len__(self):
        """
        :return: dataset size
        """

    def make_tf_dataset(self):
        """
        Make a tensorflow dataset.

        This method will do the following two steps:
          1) Open a petastorm reader on the materialized dataset dir.
          2) Create a tensorflow dataset based on the reader created in (1)

        :return: a context manager for a `tf.data.Dataset` object.
                 when exit the returned context manager, the reader
                 will be closed.
        """

    def make_torch_dataloader(self,
                              batch_size=32,
                              num_epochs=None,
                              workers_count=None,
                              cur_shard=None,
                              shard_count=None,
                              **petastorm_reader_kwargs):
        """
        Make a PyTorch DataLoader.

        This method will do the following two steps:
          1) Open a petastorm reader on the materialized dataset dir.
          2) Create a PyTorch DataLoader based on the reader created in (1)

        :param batch_size: The number of items to return per batch
        :param num_epochs: An epoch is a single pass over all rows in the
            dataset. Setting ``num_epochs`` to ``None`` will result in an
            infinite number of epochs.
        :param workers_count: An int for the number of workers to use in the
            reader pool. This only is used for the thread or process pool.
            Defaults to None, which means using the default value from
            `petastorm.make_batch_reader()`.
        :param cur_shard: An int denoting the current shard number. Each node
            reading a shard should pass in a unique shard number in the range
            [0, shard_count). shard_count must be supplied as well. Defaults to
            None
        :param shard_count: An int denoting the number of shards to break this
            dataset into. Defaults to None
        :param petastorm_reader_kwargs: all the arguments for
            `petastorm.make_batch_reader()`.

        :return: a context manager for a `torch.utils.data.DataLoader` object.
                 when exit the returned context manager, the reader
                 will be closed.
        """

    def delete(self):
        """
        Delete cache files at self.cache_dir_url.
        """
```
# Example Code (PyTorch)
```python
from petastorm import make_spark_converter
from petastorm.spark import SparkDatasetConverter
import torch

# specify a cache dir first.
# the dir is used to save materialized spark dataframe files
spark.conf.set(SparkDatasetConverter.PARENT_CACHE_DIR_URL_CONF, 'hdfs:/...')

df1 = ... # `df1` is a spark dataframe

# create a converter from `df1`
# it will materialize `df1` to cache dir.
converter1 = make_spark_converter(df1)

# make a tensorflow dataset from `converter1
with converter1.make_torch_dataloader() as dataloader:
    # the `dataloader` is `torch.utils.data.DataLoader` object
    # we can train/evaluate model on `dataloader`
    # when exit the with context, the reader of the dataloader will be closed
    ...

# delete the cached files of the dataframe.
converter1.delete()
```

### CODE REVIEW

File: petastorm/spark/spark_dataset_converter.py [hunk 1]
```diff
@@ -13,18 +13,18 @@
...
 import uuid
-import logging
 import warnings
 
 from pyarrow import LocalFileSystem
 from pyspark.sql.session import SparkSession
 from six.moves.urllib.parse import urlparse
 
-from petastorm import make_batch_reader
+import petastorm
```
Comment (mengxr) [inline petastorm/spark/spark_dataset_converter.py]: check if we can patch make_batch_reader in pytest

File: petastorm/spark/spark_dataset_converter.py [hunk 2]
```diff
@@ -153,6 +153,49 @@ def make_tf_dataset(self):
         """
         return TFDatasetContextManager(self.cache_dir_url)
 
+    def make_torch_dataloader(self,
+                              batch_size=32,
+                              num_epochs=None,
+                              workers_count=None,
+                              cur_shard=None,
```
Comment (mengxr) [inline petastorm/spark/spark_dataset_converter.py]: we can pass cur_shard shard_count to petastorm_reader

File: petastorm/spark/spark_dataset_converter.py [hunk 3]
```diff
@@ -183,6 +226,38 @@ def __exit__(self, exc_type, exc_value, exc_traceback):
...
+        :param data_url: A string specifying the data URL.
+        See `SparkDatasetConverter.make_torch_dataloader()` for the definitions
+        of the other parameters.
+        """
+        from petastorm.pytorch import DataLoader
+
+        petastorm_reader_kwargs["num_epochs"] = num_epochs
+        if workers_count is not None:
+            petastorm_reader_kwargs["workers_count"] = workers_count
+        petastorm_reader_kwargs["cur_shard"] = cur_shard
```
Comment (mengxr) [inline petastorm/spark/spark_dataset_converter.py]: need a util function to check whether this is running as an MPI environment, warn users if MPI_COMM_SIZE is not 1.

File: petastorm/spark/spark_dataset_converter.py [hunk 4]
```diff
@@ -183,6 +226,38 @@ def __exit__(self, exc_type, exc_value, exc_traceback):
...
+        petastorm_reader_kwargs["shard_count"] = shard_count
+
+        self.reader = petastorm.make_batch_reader(data_url, **petastorm_reader_kwargs)
```
Comment (mengxr) [inline petastorm/spark/spark_dataset_converter.py]: move it to __enter__

File: petastorm/tests/test_spark_dataset_converter.py [hunk 5]
```diff
@@ -228,12 +231,167 @@ def map_fn(_):
...
+            "cache_location": cache_location,
+            "cache_size_limit": cache_size_limit,
+            "cache_row_size_estimate": cache_row_size_estimate,
+            "cache_extra_settings": cache_extra_settings,
+            "hdfs_driver": hdfs_driver,
+            "transform_spec": transform_spec,
+        }
+
+    original_fn = petastorm.make_batch_reader
+    petastorm.make_batch_reader = mock_make_batch_reader
```
Comment (mengxr) [inline petastorm/tests/test_spark_dataset_converter.py]: wrap the real make_batch_reader but get the input params there

File: petastorm/tests/test_spark_dataset_converter.py [hunk 6]
```diff
@@ -228,12 +231,167 @@ def map_fn(_):
...
+    assert ctm.reader["cur_shard"] == "shard_1"
+    assert ctm.reader["shard_count"] == "total_shard"
+    assert ctm.reader["cache_type"] == "cache_1"
+    assert ctm.reader["cache_location"] == "location_1"
+    assert ctm.reader["cache_size_limit"] == "limit_1"
+    assert ctm.reader["cache_extra_settings"] == "extra_1"
+    assert ctm.reader["hdfs_driver"] == "driver_1"
+    assert ctm.reader["transform_spec"] == "transform_spec_1"
+
+    petastorm.make_batch_reader = original_fn
```
Comment (mengxr) [inline petastorm/tests/test_spark_dataset_converter.py]: need a contextmanager here to ensure we put it back

File: petastorm/tests/test_spark_dataset_converter.py [hunk 7]
```diff
@@ -228,12 +231,167 @@ def map_fn(_):
...
+                               predicate=None,
+                               rowgroup_selector=None,
+                               num_epochs=1,
+                               cur_shard=None, shard_count=None,
+                               cache_type='null', cache_location=None, cache_size_limit=None,
+                               cache_row_size_estimate=None, cache_extra_settings=None,
+                               hdfs_driver='libhdfs3',
+                               transform_spec=None):
+        return {
+            "dataset_url": dataset_url,
```
Comment (mengxr) [inline petastorm/tests/test_spark_dataset_converter.py]: it should be okay if you only check 1 or 2 to keep the test simple

**Comment by WeichenXu123** [comment]
To mock `make_batch_reader` for test, the following simple code should work:
~~~python
from contextlib import contextmanager
@contextmanager
def mock_make_batch_reader():
	captured_args = []
	import petastorm
	original_make_batch_reader = petastorm.make_batch_reader
	def mock_fn(dataset_url, **kwargs):
		captured_args.append({'dataset_url': dataset_url, **kwargs})
		return original_make_batch_reader(dataset_url, **kwargs)
	petastorm.make_batch_reader = mock_fn
	try:
		yield captured_args
	finally:
		petastorm.make_batch_reader = original_make_batch_reader
with mock_make_batch_reader() as captured_args:
	from petastorm import make_batch_reader
	with make_batch_reader('file:///tmp/t0001', workers_count=18) as reader:
		for i in reader:
    			print(i)
	print('get captured args: ' + str(captured_args))
~~~

File: petastorm/spark/spark_dataset_converter.py [hunk 8]
```diff
@@ -153,6 +169,49 @@ def make_tf_dataset(self):
         """
         return TFDatasetContextManager(self.cache_dir_url)
 
+    def make_torch_dataloader(self,
+                              batch_size=32,
+                              num_epochs=None,
+                              workers_count=4,
```
Comment (mengxr) [inline petastorm/spark/spark_dataset_converter.py]: We might want to leave the default to None so we can autotune it in the future.

File: petastorm/tests/test_spark_dataset_converter.py [hunk 9]
```diff
@@ -222,18 +228,160 @@ def map_fn(_):
...
+    df = test_ctx.spark.range(8)
+    conv = make_spark_converter(df)
+
+    with pytest.raises(TypeError, match="unexpected keyword argument 'xyz'"):
+        with conv.make_torch_dataloader(xyz=1) as _:
+            pass
+
+
+@contextmanager
+def mock_torch_make_batch_reader():
```
Comment (selitvin) [inline petastorm/tests/test_spark_dataset_converter.py]: Would it be simpler to use `mock.patch`?

File: petastorm/spark/spark_dataset_converter.py [hunk 10]
```diff
@@ -153,6 +169,49 @@ def make_tf_dataset(self):
...
+        :return: a context manager for a `torch.utils.data.DataLoader` object.
+                 when exit the returned context manager, the reader
+                 will be closed.
+        """
+
+        env_rank, env_size = get_env_rank_and_size()
+        if env_rank is not None and env_size is not None:
+            if petastorm_reader_kwargs['cur_shard'] != env_rank or \
+                    petastorm_reader_kwargs['shard_count'] != env_size:
+                warnings.warn(
```
Comment (selitvin) [inline petastorm/spark/spark_dataset_converter.py]: Let's cover this rank consistency logic in a test.

**Comment by WeichenXu123** [comment]
Let's wait this PR https://github.com/uber/petastorm/pull/506 merge first,
and then reuse some code there.

**Comment by praateekmahajan** [comment]
@liangz1 @mengxr  @selitvin do we know if this PR can leverage the performance boost being offered in #492? If yes, it might be a nice idea to get that merged too, given all tests pass.

**Comment by WeichenXu123** [comment]
Looks Good!

**Pull request merged**